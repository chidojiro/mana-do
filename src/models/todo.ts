export type TodoStatus = 'ACTIVE' | 'COMPLETED';

export interface Todo {
  id: string;
  user_id: string;
  content: string;
  status?: TodoStatus;
  created_date: string;
}
