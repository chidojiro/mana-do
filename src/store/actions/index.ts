import { TodoActions as TodoActionsType } from './todos.action';

export {
  deleteAllTodos,
  deleteTodo,
  setTodos,
  updateAllTodosStatus,
  updateTodoStatus,
  updateTodoContent,
} from './todos.action';

export type TodoActions = TodoActionsType;
