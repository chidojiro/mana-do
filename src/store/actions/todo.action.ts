import { Todo } from '../../models/todo';

// GET_TODO
export interface SetTodoAction {
  type: 'GET_TODOS';
}
export function getTodos(): SetTodoAction {
  return {
    type: 'GET_TODOS',
  };
}

// CREATE_TODO
export interface CreateTodoAction {
  type: 'CREATE_TODO';
  payload: Todo;
}
export function createTodo(newTodo: Todo): CreateTodoAction {
  return {
    type: 'CREATE_TODO',
    payload: newTodo,
  };
}

// UPDATE_TODO_STATUS
export interface UpdateTodoStatusAction {
  type: 'UPDATE_TODO_STATUS';
  payload: {
    todoId: string;
    checked: boolean;
  };
}
export function updateTodoStatus(todoId: string, checked: boolean): UpdateTodoStatusAction {
  return {
    type: 'UPDATE_TODO_STATUS',
    payload: {
      todoId,
      checked,
    },
  };
}

// DELETE_TODO
export interface DeleteTodoAction {
  type: 'DELETE_TODO';
  payload: string;
}
export function deleteTodo(todoId: string): DeleteTodoAction {
  return {
    type: 'DELETE_TODO',
    payload: todoId,
  };
}

// DELETE_ALL_TODOS
export interface DeleteAllTodosAction {
  type: 'DELETE_ALL_TODOS';
}
export function deleteAllTodos(): DeleteAllTodosAction {
  return {
    type: 'DELETE_ALL_TODOS',
  };
}

// TOGGLE_ALL_TODOS
export interface ToggleAllTodosAction {
  type: 'TOGGLE_ALL_TODOS';
  payload: boolean;
}
export function toggleAllTodos(checked: boolean): ToggleAllTodosAction {
  return {
    type: 'TOGGLE_ALL_TODOS',
    payload: checked,
  };
}

export type TodoActions =
  | SetTodoAction
  | CreateTodoAction
  | UpdateTodoStatusAction
  | DeleteTodoAction
  | DeleteAllTodosAction
  | ToggleAllTodosAction;
