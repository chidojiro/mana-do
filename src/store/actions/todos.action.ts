import { Dispatch } from 'redux';
import {
  getTodos as getTodosApi,
  createTodo as createTodoApi,
  updateTodoStatus as updateTodoStatusApi,
  deleteTodo as deleteTodoApi,
  updateAllTodosStatus as updateAllTodosStatusApi,
  deleteAllTodos as deleteAllTodosApi,
  updateTodoContent as updateTodoContentApi,
} from '../../apis/todo.apis';
import { TodoStatus } from '../../models/todo';
import { Todo } from '../../models/todo';

// SET_TODO
export interface SetTodoAction {
  type: 'SET_TODOS';
  payload: Todo[];
}
export function setTodos(todo: Todo[]): SetTodoAction {
  return {
    type: 'SET_TODOS',
    payload: todo,
  };
}

// GET_TODO
export const getTodos = () => async (dispatch: Dispatch) => {
  const todos = await getTodosApi();

  dispatch({
    type: 'SET_TODOS',
    payload: todos,
  });
};

// CREATE_TODO
export interface CreateTodoAction {
  type: 'CREATE_TODO';
  payload: Todo;
}
export const createTodo = (newTodoContent: string) => async (dispatch: Dispatch) => {
  const newTodo = await createTodoApi(newTodoContent);

  dispatch({
    type: 'CREATE_TODO',
    payload: newTodo,
  });
};

// UPDATE_TODO_STATUS
export interface UpdateTodoStatusAction {
  type: 'UPDATE_TODO_STATUS';
  payload: {
    todoId: string;
    status: TodoStatus;
  };
}
export const updateTodoStatus = (todoId: string, status: TodoStatus) => async (dispatch: Dispatch) => {
  await updateTodoStatusApi(todoId, status);

  dispatch({
    type: 'UPDATE_TODO_STATUS',
    payload: {
      todoId,
      status,
    },
  });
};

// DELETE_TODO
export interface DeleteTodoAction {
  type: 'DELETE_TODO';
  payload: string;
}
export const deleteTodo = (todoId: string) => async (dispatch: Dispatch) => {
  await deleteTodoApi(todoId);

  dispatch({
    type: 'DELETE_TODO',
    payload: todoId,
  });
};

// DELETE_ALL_TODOS
export interface DeleteAllTodosAction {
  type: 'DELETE_ALL_TODOS';
}
export const deleteAllTodos = () => async (dispatch) => {
  await deleteAllTodosApi();

  dispatch({
    type: 'DELETE_ALL_TODOS',
  });
};

// TOGGLE_ALL_TODOS
export interface UpdateAllTodosStatusAction {
  type: 'UPDATE_ALL_TODOS_STATUS';
  payload: TodoStatus;
}
export const updateAllTodosStatus = (status: TodoStatus) => async (dispatch: Dispatch) => {
  await updateAllTodosStatusApi(status);

  dispatch({
    type: 'UPDATE_ALL_TODOS_STATUS',
    payload: status,
  });
};

export interface UpdateTodoContent {
  type: 'UPDATE_TODO_CONTENT';
  payload: {
    todoId: string;
    content: string;
  };
}
export const updateTodoContent = (todoId: string, content: string) => async (dispatch: Dispatch) => {
  await updateTodoContentApi(todoId, content);

  dispatch({ type: 'UPDATE_TODO_CONTENT', payload: { todoId, content } });
};

export type TodoActions =
  | SetTodoAction
  | CreateTodoAction
  | UpdateTodoStatusAction
  | DeleteTodoAction
  | DeleteAllTodosAction
  | UpdateAllTodosStatusAction
  | UpdateTodoContent;
