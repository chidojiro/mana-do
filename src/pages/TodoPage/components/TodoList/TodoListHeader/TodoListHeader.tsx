import React from 'react';

import { Switch } from '../../../../../components';
import { useSelector, useDispatch } from 'react-redux';
import { todosSelector } from '../../../../../store/reducers/todos.reducer';
import { updateAllTodosStatus, deleteAllTodos } from '../../../../../store/actions';

export const TodoListHeader = () => {
  const dispatch = useDispatch();
  const todos = useSelector(todosSelector);

  const isAllTodosCompleted = todos.every(({ status }) => status === 'COMPLETED');

  const handleToggleAllStatus = (value) => {
    if (value) {
      dispatch(updateAllTodosStatus('COMPLETED'));
    } else {
      dispatch(updateAllTodosStatus('ACTIVE'));
    }
  };

  const handleDeleteClick = () => {
    dispatch(deleteAllTodos());
  };

  return (
    <div className='todo-list__row todo-list__header'>
      <div className='todo-details'>TODO</div>
      <div className='todo-status' title='Toggle all todos'>
        <Switch value={isAllTodosCompleted} onChange={handleToggleAllStatus} />
      </div>
      <div className='todo-actions'>
        <i className='fas fa-trash todo-card__delete-icon' onClick={handleDeleteClick} title='Delete all'></i>
      </div>
    </div>
  );
};
