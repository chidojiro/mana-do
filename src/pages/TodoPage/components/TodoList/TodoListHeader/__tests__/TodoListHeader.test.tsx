jest.mock('react-redux', () => ({
  useSelector: jest.fn(),
  useDispatch: jest.fn(),
}));
jest.mock('../../../../../../store/actions', () => ({ updateAllTodosStatus: jest.fn(), deleteAllTodos: jest.fn() }));

import React from 'react';
import { shallow } from 'enzyme';
import toJSON from 'enzyme-to-json';
import { useSelector, useDispatch } from 'react-redux';

import { TodoListHeader } from '../TodoListHeader';
import { Store } from '../../../../../../store/reducers/root.reducer';
import { Todo, TodoStatus } from '../../../../../../models';
import { updateAllTodosStatus, deleteAllTodos } from '../../../../../../store/actions';

describe('TodoListHeader', () => {
  const completedTodos = [{ status: 'COMPLETED' } as Todo, { status: 'COMPLETED' } as Todo];

  const incompletedTodos = [{ status: 'COMPLETED' } as Todo, { status: 'ACTIVE' } as Todo];

  const defaultStore: Store = {
    todos: { items: incompletedTodos },
  };

  const mockDispatch = jest.fn();

  const updateAllTodosStatusAction = {
    type: 'updateAllTodosStatus',
  };

  const deleteAllTodosAction = {
    type: 'deleteAllTodos',
  };

  beforeAll(() => {
    (useSelector as jest.Mock).mockImplementation((selector) => selector(defaultStore));
    (useDispatch as jest.Mock).mockImplementation(() => mockDispatch);

    (updateAllTodosStatus as jest.Mock).mockReturnValue(updateAllTodosStatusAction);
    (deleteAllTodos as jest.Mock).mockReturnValue(deleteAllTodosAction);
  });

  beforeEach(() => {
    jest.clearAllMocks();
  });

  it('should render the component', () => {
    const wrapper = shallow(<TodoListHeader />);

    expect(toJSON(wrapper)).toMatchSnapshot();
  });

  it('should activate the switch when all todos is completed', () => {
    const mockStore: Store = {
      todos: { items: completedTodos },
    };
    (useSelector as jest.Mock).mockImplementation((selector) => selector(mockStore));

    const wrapper = shallow(<TodoListHeader />);

    expect(wrapper.find('Switch').prop('value')).toBe(true);
  });

  it('should deactivate the switch when any of the todos is not completed', () => {
    (useSelector as jest.Mock).mockImplementation((selector) => selector(defaultStore));

    const wrapper = shallow(<TodoListHeader />);

    expect(wrapper.find('Switch').prop('value')).toBe(false);
  });

  it('should mark all todos as active when the switch is clicked at completed state', () => {
    const mockStore: Store = {
      todos: { items: completedTodos },
    };
    (useSelector as jest.Mock).mockImplementation((selector) => selector(mockStore));

    const mockDispatch = useDispatch();

    const wrapper = shallow(<TodoListHeader />);

    wrapper.find('Switch').simulate('change', false);

    expect(updateAllTodosStatus).toBeCalledWith('ACTIVE');
    expect(mockDispatch).toBeCalledWith(updateAllTodosStatusAction);
  });

  it('should mark all todos as completed when the switch is clicked at active state', () => {
    const mockStore: Store = {
      todos: { items: completedTodos },
    };
    (useSelector as jest.Mock).mockImplementation((selector) => selector(mockStore));

    const mockDispatch = useDispatch();

    const wrapper = shallow(<TodoListHeader />);

    wrapper.find('Switch').simulate('change', true);

    expect(updateAllTodosStatus).toBeCalledWith('COMPLETED');
    expect(mockDispatch).toBeCalledWith(updateAllTodosStatusAction);
  });

  it('should delete all todos when delete button is clicked', () => {
    const wrapper = shallow(<TodoListHeader />);

    wrapper.find('.todo-card__delete-icon').simulate('click');

    expect(deleteAllTodos).toBeCalled();
    expect(mockDispatch).toBeCalledWith(deleteAllTodosAction);
  });
});
