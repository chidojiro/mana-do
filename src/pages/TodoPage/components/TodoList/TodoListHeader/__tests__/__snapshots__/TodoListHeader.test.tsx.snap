// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`TodoListHeader should render the component 1`] = `
<div
  className="todo-list__row todo-list__header"
>
  <div
    className="todo-details"
  >
    TODO
  </div>
  <div
    className="todo-status"
    title="Toggle all todos"
  >
    <Switch
      onChange={[Function]}
      value={false}
    />
  </div>
  <div
    className="todo-actions"
  >
    <i
      className="fas fa-trash todo-card__delete-icon"
      onClick={[Function]}
      title="Delete all"
    />
  </div>
</div>
`;
